PDBs: mention scale subresource in Arbitrary Controllers and Selectors (#39196)
kubernetes/website: content/en/docs/tasks/run-application/configure-pdb.md

> ## Arbitrary Controllers and Selectors

Reviewer: ```suggestion
## Arbitrary object kinds and arbitrary selectors {#arbitrary-controllers-and-selectors}
```

Author: As mentioned in https://github.com/kubernetes/website/pull/39196#discussion_r1100583119 I would prefer we use something like Workload resources. But in the title I would rather say `workloads` since it applies transitively to pods as well. 

Object kinds would also suggest that the PDB is managing other kinds and not just pods.

Also, do we want to duplicate the arbitrary word? Does it help with comprehensiveness/ambiguity?

You can skip this section if you only use PDBs with the built-in
application controllers (Deployment, ReplicationController, ReplicaSet, and StatefulSet)

> or controllers that implement a [scale subresource](/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#scale-subresource),

Author: not sure if we should have the link 2nd time since it is linked already in the beginning of this article

> You can skip this section if you only use PDBs with the built-in
> application controllers (Deployment, ReplicationController, ReplicaSet, and StatefulSet)
> or controllers that implement a [scale subresource](/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#scale-subresource),

Reviewer: ```suggestion
You can skip this section if you only use PDBs with the built-in
workload management object kinds (Deployment, ReplicaSet, StatefulSet and ReplicationController),
or with {{< glossary_tooltip term_id="CustomResourceDefinition" text="custom resources" >}} that
implement a `scale` [subresource](/docs/concepts/extend-kubernetes/api-extension/custom-resources/#advanced-features-and-flexibility).
```

We tend not to call Deployment a controller. One reason is that if you call something ReplicationController, and then you have a control loop that takes the `.spec` of a ReplicationController as its desired state, you then kind of have to call the code the “ReplicationController controller”.

I'd prefer to revise the wording whilst we're touching this section. Also, I propose adding a glossary tooltip to explain CRD.

Author: I guess it does not make that much sense to reference it as a controller from user's perspective here.

Workload management object kinds sounds too technical/specific. Can we call them workload resources? (we use this term in the concepts as well https://kubernetes.io/docs/concepts/workloads/controllers/)

I made the changes to the rest of this section. But we mention controller in the rest of the document. So should we rewrite it all, or do a followup with a focus on that?

+1 for the glossary.

Regarding the scale subresource link - the proposed one gives a high level definition which links to the more detailed one. Not sure which one is better here.

Nevertheless we probably want to be consistent with the link in the intro, right? Should I swap it as well?

Author: I am going back and forth regarding the controller vs workload resource.

The pod should be owned by a resource which is a controller ([ownerReferences.controller](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/object-meta/#System)), so from k8s perspective the ReplicaSet is the controller and I am not sure if we want to diverge from that.

> with the PDB selector matching the controller's selector.

Author: the whole sentence is becoming convoluted a bit. Should I itemize it?

Reviewer 2: from my side it looks good

> with the PDB selector matching the controller's selector.

Reviewer: ```suggestion
and where the PDB selector exactly matches the selector of the object that holds the Pod template.
```

Author: I would rather not mention the pod template, because the owning resource might not have it and could just generate the template on the fly.

Pods in `Pending`, `Succeeded` or `Failed` phase are always considered for eviction.
{{< /note >}}

## Arbitrary workloads and arbitrary selectors {#arbitrary-controllers-and-selectors}

You can skip this section if you only use PDBs with the built-in
workload resources (Deployment, ReplicaSet, StatefulSet and ReplicationController)
or with {{< glossary_tooltip term_id="CustomResourceDefinition" text="custom resources" >}}
that implement a `scale` [subresource](/docs/concepts/extend-kubernetes/api-extension/custom-resources/#advanced-features-and-flexibility),
and where the PDB selector exactly matches the selector of the Pod's owning resource.

You can use a PDB with pods controlled by another resource, by an
"operator", or bare pods, but with these restrictions:

- only `.spec.minAvailable` can be used, not `.spec.maxUnavailable`.

> - only an integer value can be used with `.spec.minAvailable`, not a percentage.

Reviewer 3: Nit, can we explain why in more details?

Author: @soltysh added the explanation for the restrictions

Pods in `Pending`, `Succeeded` or `Failed` phase are always considered for eviction.
{{< /note >}}

## Arbitrary workloads and arbitrary selectors {#arbitrary-controllers-and-selectors}

You can skip this section if you only use PDBs with the built-in
workload resources (Deployment, ReplicaSet, StatefulSet and ReplicationController)
or with {{< glossary_tooltip term_id="CustomResourceDefinition" text="custom resources" >}}
that implement a `scale` [subresource](/docs/concepts/extend-kubernetes/api-extension/custom-resources/#advanced-features-and-flexibility),
and where the PDB selector exactly matches the selector of the Pod's owning resource.

You can use a PDB with pods controlled by another resource, by an
"operator", or bare pods, but with these restrictions:

- only `.spec.minAvailable` can be used, not `.spec.maxUnavailable`.
- only an integer value can be used with `.spec.minAvailable`, not a percentage.

It is not possible to use other availability configurations,

> because we cannot derive a total number of pods without a supported owning resource.

Reviewer: (nit)
```suggestion
because Kubernetes cannot derive a total number of pods without a supported owning resource.
```

Author: @sftim thanks for noticing that. Fixed.
